proposals/explicit-api-mode.md

```diff
@@ -0,0 +1,154 @@
+
+## Implementation
+
+This mode is enabled by the compiler flag `-Xexplicit-api={strict|warning}`. `strict` state of the flag means that compiler will issue errors when public API declaration does not have explicit visibility or explicit return type.
+`warning` means that the compiler will issue warnings (this will help migration).
+Note: missing KDoc is always a warning, regardless of the state of the flag.
+
+To ease setting up this mode, a DSL would be provided in Kotlin Gradle plugin:
+
+```gradle
+kotlinOptions {
+    explicitApi()
+    // or
+    explicitApi = ExplicitApiMode.Strict
+    // or
+    explicitApiWarning()
+    // or
+    explicitApi = ExplicitApiMode.Warning
+}
+```
+Explicit mode enabled here would not affect test sources.
+
+The exact place of these methods (top-level `kotlinOptions`, or compilation, or particular source set) is TBA. Maven plugin option is TBD.
+
+Kotlin plugin in IntelliJ IDEA would recognize that explicit mode is enabled and would offer corresponding intentions and quick-fixes.
+
+The following **IDE** inspections would be disabled either because they're redundant or duplicated in the compiler:
+
+* 'Redundant visibility modifier' for public API declarations
+* 'Public API declaration has implicit return type' (would be replaced with compiler diagnostic)
+* 'Missing KDoc comment for public declaration' (would be replaced with compiler diagnostic)
+
+Effectively, 'disabled in IDE/replaced with compiler diagnostic' means that you would not be able to control these inspections via 'Preferences - Editor - Inspections'.
+
+## Alternatives
+
+Besides embedding such mode as a compiler flag, the following alternatives were considered and discarded:
+
+1. Compiler plugin
+
+    While it is possible to make this functionality as a separate plugin, the lack of stability and consistency in the compiler plugin API (because there is no official public definition of plugin API/compiler surface, actually) would require a lot of efforts to support this relatively small feature.
+
+2. External tool
+
+    An external tool like a style checker may look like a proper solution for this problem, but it requires much more additional setup than one DSL parameter, which pushes adoption back. Another big problem is that such tools re-analyze whole code one more time, which leads to increased build times, big size of the tool, or incorrect results.
+
+## Compatibility
+
+The explicit mode doesn't change the semantics of the correct code and does not affect bytecode generation, so it's safe from the standpoint of compatibility.
+For example, if some code was compiled once in the explicit mode and later without explicit mode, the result should be the same.
+
+However, this mode could make previously correct code to compile with errors (the code that had the default visibility declarations and inferred types in public API).
```

Comment:
"could make previously correct code ~to~ compile with errors"  (remove "_to_" here)